src/swarm/neo/request/RequestEventDispatcher.d

```diff
@@ -0,0 +1,777 @@
+            this.dispatchSignal(conn, code);
+        }
+    }
+
+    /***************************************************************************
+
+        Helper function for eventLoop(). Dispatches the specified signal to the
+        appropriate waiting fiber.
+
+        Params:
+            conn = request-on-conn event dispatcher (used to throw a protocol
+                error, if the signal is not handled)
+            signal = signal code
+
+        Throws:
+            if the signal is not handled by any waiting fiber
+
+    ***************************************************************************/
+
+    private void dispatchSignal ( RequestOnConnBase.EventDispatcher conn,
+        int signal )
+    {
+        if ( signal < 0 ) // Standard send/receive codes
+            return;
+
+        assert(signal <= 255);
+        ubyte signal_ubyte = cast(ubyte)signal;
+
+        foreach ( waiting_fiber; this.waiting_fibers )
+        {
+            if ( waiting_fiber.event.active == waiting_fiber.event.active.signal
+                && waiting_fiber.event.signal.code == signal_ubyte )
+            {
+                EventNotification fired_event;
+                fired_event.signal = Signalled(signal_ubyte);
+
+                this.notifyWaitingFiber(waiting_fiber.fiber, fired_event);
+                return;
+            }
+        }
+
+        log.warn("Unhandled signal of type {}", signal_ubyte);
+        conn.shutdownWithProtocolError("Unhandled signal code");
+    }
+
+    /***************************************************************************
+
+        Helper function for eventLoop(). Dispatches the specified message
+        payload to the appropriate waiting fiber, via the first byte which
+        represents the type of the message.
+
+        Params:
+            conn = request-on-conn event dispatcher (used to parse the message
+                type from the payload and to throw a protocol error, if the
+                message type is not handled)
+            payload = received message
+
+        Throws:
+            if the message type is not handled by any waiting fiber
+
+    ***************************************************************************/
+
+    private void dispatchReceivedPayload ( RequestOnConnBase.EventDispatcher conn,
+        Const!(void)[] payload )
+    {
+        auto message_type = *conn.message_parser.getValue!(ubyte)(payload);
+
+        foreach ( waiting_fiber; this.waiting_fibers )
+        {
+            if ( waiting_fiber.event.active == waiting_fiber.event.active.message
+                && waiting_fiber.event.message.type == message_type )
+            {
+                EventNotification fired_event;
+                fired_event.message = ReceivedMessage(message_type, payload);
+
+                this.notifyWaitingFiber(waiting_fiber.fiber, fired_event);
+                return;
+            }
+        }
+
+        log.warn("Unhandled message of type {}", message_type);
```

Comment:
Hm, it's a tricky point. I deliberately made this a relatively high log level as it really should be noticed if this ever happens. (I'd envisaged using this in combination with the log level stats writer in ocean.)

I see what you mean, though. If the time between the buggy request starting, this error occurring, and the client firing the request again was very small, it could lead to a big performance hit affecting other requests.